iree/compiler/Dialect/Flow/IR/FlowOpFolders.cpp

```diff
@@ -695,13 +695,35 @@ struct FlattenTensorReshapeChain : public OpRewritePattern<TensorReshapeOp> {
   }
 };
 
+// Replace `flow.tensor.splat`-`flow.tensor.load` op-pairs by the input
+// primitive value for the splat op.
+struct FoldSplatLoadIntoPrimitive : public OpRewritePattern<TensorLoadOp> {
+  using OpRewritePattern<TensorLoadOp>::OpRewritePattern;
+
+  LogicalResult matchAndRewrite(TensorLoadOp loadOp,
+                                PatternRewriter &rewriter) const override {
+    auto sourceOp =
```

Comment:
out of curiosity when looking at real inputs with flow control are the splat and load often in the same basic block? (this will fail for block arguments, which is fine, but just wondering if we'll need to prioritize doing iter-block code motion to try to get these together)

Reply:
If I understand the dialect conversion framework properly then yes. This is because:
- The `load` op is actually part of the input IR, i.e., it isn't a result of the detensoring pass.
- The tensor from which `load` reads is a result of a `source materialization` that takes place during detensoring as setup by the dialect conversion framework. Specifically here: https://github.com/llvm/llvm-project/blob/c691651c53485a07db5b529e5a465eee397b7987/mlir/lib/Dialect/Linalg/Transforms/Detensorize.cpp#L155.

If I understand correctly, then this materialization should happen locally where ever the original/detensored value was used.